src/platform/EFR32/KeyValueStoreManagerImpl.h

```diff
@@ -0,0 +1,214 @@
+#include <pw_kvs/key_value_store.h>
+
+#if defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
+
+// KVS is only available for EFR32 when these macros are defined.
+#if defined(CHIP_KVS_SECTOR_COUNT) && defined(CHIP_KVS_BASE_SECTOR_INDEX)
+#define CHIP_KVS_AVAILABLE 1
+#else // defined(CHIP_KVS_SECTOR_COUNT) && defined(CHIP_KVS_BASE_ADDRESS)
+#define CHIP_KVS_AVAILABLE 0
+#endif // defined(CHIP_KVS_SECTOR_COUNT) && defined(CHIP_KVS_BASE_ADDRESS)
+
+namespace chip {
+namespace DeviceLayer {
+namespace PersistedStorage {
+
+#if defined(CHIP_KVS_AVAILABLE) && CHIP_KVS_AVAILABLE
+class KeyValueStoreManagerImpl final : public KeyValueStoreManager
+{
+    // Allow the KeyValueStoreManager interface class to delegate method calls to
+    // the implementation methods provided by this class.
+    friend class KeyValueStoreManager;
+
+public:
+    void Init() { mKvs.Init(); }
+
+    CHIP_ERROR _Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size = nullptr, size_t offset = 0) const;
+
+    CHIP_ERROR _Delete(const char * key);
+
+    /**
+     * @brief
+     * Erases all data in the KVS partition, KVS needs to be initalized after
+     * this operation.
+     *
+     * @return CHIP_NO_ERROR the partiton was erased.
+     *         CHIP_ERROR_TIMEOUT timed out while doing erase.
+     *         CHIP_ERROR_ACCESS_DENIED flash locked, erase failed.
+     */
+    CHIP_ERROR ErasePartition();
+
+    CHIP_ERROR _Put(const char * key, const void * value, size_t value_size);
+
+private:
+    // KVS flash interface
+    class Efr32FlashMemory : public pw::kvs::FlashMemory
+    {
+    public:
+        Efr32FlashMemory() : pw::kvs::FlashMemory(FLASH_PAGE_SIZE, FLASH_SIZE / FLASH_PAGE_SIZE, sizeof(uint32_t), FLASH_BASE) {}
+
+        // Enabling flash handled by platform
+        pw::Status Enable() override { return pw::Status::OK; }
+        pw::Status Disable() override { return pw::Status::OK; }
+        bool IsEnabled() const override { return true; }
+
+        pw::Status Erase(Address flash_address, size_t num_sectors) override
+        {
+            for (size_t i = 0; i < num_sectors; i++)
+            {
+                auto status =
+                    MscStatusToPwStatus(MSC_ErasePage(reinterpret_cast<uint32_t *>(flash_address + i * sector_size_bytes())));
+                if (!status.ok())
+                {
+                    return status;
+                }
+            }
+            return pw::Status::OK;
+        }
+
+        pw::StatusWithSize Read(Address address, std::span<std::byte> output) override
+        {
+            memcpy(output.data(), reinterpret_cast<void *>(address), output.size());
+            return pw::StatusWithSize::Ok(output.size());
+        }
+
+        pw::StatusWithSize Write(Address destination_flash_address, std::span<const std::byte> data) override
+        {
+            return pw::StatusWithSize(MscStatusToPwStatus(MSC_WriteWord(reinterpret_cast<uint32_t *>(destination_flash_address),
```

Comment:
This `reinterpret_cast` looks like undefined behavior. What if I pass in a span that's not 32 bit aligned?

Reply:
This interface is private and only available to the KVS object which is provided the alignment. I'll add some asserts though just in case.